# <a target="_blank" href="https://miptstats.github.io/courses/ad_mipt.html">Phystech@DataScience</a>


In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns
sns.set(font_scale=1.3)

### Линейная регрессия: практика
Перед началом работы догрузим необходимые нам функции



In [2]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

Рассмотрим данные исследования 2004 года о зависимости риска развития диабета от пола, возраста, индекса массы тела, среднего кровяного давления и других показателей. 



#### Загрузка данных

In [3]:
data = load_diabetes()
data.data

array([[ 0.03807591,  0.05068012,  0.06169621, ..., -0.00259226,
         0.01990749, -0.01764613],
       [-0.00188202, -0.04464164, -0.05147406, ..., -0.03949338,
        -0.06833155, -0.09220405],
       [ 0.08529891,  0.05068012,  0.04445121, ..., -0.00259226,
         0.00286131, -0.02593034],
       ...,
       [ 0.04170844,  0.05068012, -0.01590626, ..., -0.01107952,
        -0.04688253,  0.01549073],
       [-0.04547248, -0.04464164,  0.03906215, ...,  0.02655962,
         0.04452873, -0.02593034],
       [-0.04547248, -0.04464164, -0.0730303 , ..., -0.03949338,
        -0.00422151,  0.00306441]])

Функция `sklearn.datasets.load_diabetes()` возвращает словарь. В поле `data` записана матрица регрессоров, в которой данные предварительно центрированы и нормированы. В поле `target` записана мера прогрессирования заболевания в течении года. В поле `DESCR` можно прочитать подробнее о данных.

Посмотрим на описание датасета.

In [4]:
print(data['DESCR'])

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

:Number of Instances: 442

:Number of Attributes: First 10 columns are numeric predictive values

:Target: Column 11 is a quantitative measure of disease progression one year after baseline

:Attribute Information:
    - age     age in years
    - sex
    - bmi     body mass index
    - bp      average blood pressure
    - s1      tc, total serum cholesterol
    - s2      ldl, low-density lipoproteins
    - s3      hdl, high-density lipoproteins
    - s4      tch, total cholesterol / HDL
    - s5      ltg, possibly log of serum triglycerides level
    - s6      glu, blood sugar level

Note: Each of these 10 feature variables have bee

Поле `data` содержит матрицу размера 442 $\times$ 10, где 442 — количество пациентов, а 10 — количество признаков (возраст, пол, и т.д.). 
Строки матрицы соответствуют пациентам, столбцы — признакам.

In [5]:
data['data'].shape

(442, 10)

Целевая переменная $-$ мера прогрессирования заболевания в течении года.

In [6]:
data['target'].shape

(442,)

Создайте матрицу регрессоров $X$ (data) и столбец наблюдений $y$ (целевая переменная).

In [7]:
X, y = (data['data'], data['target'])

#### Обучение моделей

Разбейте данные случайно на две части — обучающую и тестовую в соотношении 80:20.

Если что-то забыли или что-то не понятно, можно почитать справку: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html

При разбиении датасета стоит зафиксировать случайность для воспроизводимости результатов, поставив `random_state=42` в функцию разбиения (можете ставить любое число, необязательно именно 42)

In [8]:
X_train, X_test = train_test_split(X, test_size=0.2, random_state = 207)
y_train, y_test = train_test_split(y, test_size=0.2, random_state = 207)

print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(353, 10) (89, 10) (353,) (89,)


Заведите модель линейной регрессии из `sklearn` и обучите ее по обучающей части данных.

In [9]:
model = LinearRegression(fit_intercept=True)  # объявляем модель
model.fit(X_train, y_train)  # обучаем

LinearRegression()

Посмотрите на результат обучения. Выведите коэффициенты перед признаками и свободный коэффициент.

In [10]:
model.coef_

array([  -7.583247  , -191.26977282,  579.85087545,  275.52512163,
       -958.09232436,  586.41587823,  200.79090086,  177.91308725,
        874.47395214,   30.92098642])

In [11]:
model.intercept_

151.03902084739215

Выполните предсказание модели на тестовой выборке.

In [12]:
y_pred = model.predict(X_test)
y_pred

array([102.39657859, 202.26765239, 203.2826672 , 152.71301996,
       244.34764801, 211.36951271,  56.23955177, 215.19659164,
       259.48330209, 256.48026621,  95.09017113, 112.77352941,
       131.98670641, 196.37663639, 213.34032688, 168.33413217,
       227.14074205, 199.28828176, 205.58716111, 148.85960017,
        33.52517415, 114.63733936,  66.1833485 ,  86.28402625,
       132.28464562, 191.15295232,  85.97422399, 162.14117213,
       141.65709138, 193.81585868, 127.861097  , 109.21727222,
       167.47508712, 149.21719447, 127.8830604 ,  78.96976344,
       189.52117022, 239.10148938,  86.19585676, 174.56224388,
       106.39359834, 254.689614  , 193.43455099, 132.30032122,
       101.97857217,  40.23678378, 100.88760167,  87.42458997,
       122.62880691, 194.23492191, 112.70303102, 198.58867273,
        59.81180212, 172.09137716, 137.617669  ,  69.87047837,
        50.39631859, 256.56247237, 210.42130736, 135.63997458,
        93.64214743,  70.83898076, 153.54608751, 163.81

Реализуйте метрики MSE, MAE, MAPE без использования `sklearn` и других готовых реализаций.

Пусть $Y_1, ..., Y_n$ &mdash; истинные значения, а $\widehat{Y}_1, ..., \widehat{Y}_n$ &mdash; предсказания.


Метрика MSE (mean squared error) определяется как
$$MSE = \frac{1}{n}\sum_{i=1}^n \left(Y_i - \widehat{Y}_i\right)^2.$$

Метрика **MAE** (*mean absolute error*), определяемая как 
$$MAE = \frac{1}{n}\sum_{i=1}^n \left|Y_i - \widehat{Y}_i\right|.$$

Метрика **MAPE** (*mean absolute percentage error*), определяемая как 
$$MAPE = 100\% \cdot \frac{1}{n}\sum_{i=1}^n \frac{\left|Y_i - \widehat{Y}_i\right|}{Y_i}.$$


In [13]:
def mean_squared_error(y_true, y_pred):
    return np.mean((y_true - y_pred)**2)

def mean_absolute_error(y_true, y_pred):
    return np.mean(np.abs(y_true - y_pred))

def mean_absolute_percentage_error(y_true, y_pred):
    return 100 * np.mean(np.abs(y_true - y_pred)/y_true)

Посчитайте MSE, MAE, MAPE на тестовой выборке и выведите с точностью до трех знаков после запятой.

In [14]:
print(np.round(mean_squared_error(y_test, y_pred), 3))
print(np.round(mean_absolute_error(y_test, y_pred), 3))
print(np.round(mean_absolute_percentage_error(y_test, y_pred), 3))



3507.363
47.969
40.609


### Различные разбиения

Разбейте данные на тренировочную и тестовую выборки еще раз: измените пераметр `random_state`. Для каждого из разбиений обучите модель и получите метрики качества. Меняются ли эти метрики, если менять разбиение данных?

In [15]:
mse = []
mae = []
mape = []

for state in [1, 495, 324, 8, 0, 900]:

    X_train, X_test = train_test_split(X, test_size=0.2, random_state = state)
    y_train, y_test = train_test_split(y, test_size=0.2, random_state = state)

    model = LinearRegression(fit_intercept=True)  # объявляем модель
    model.fit(X_train, y_train)  # обучаем


    # предсказание модели
    y_pred = model.predict(X_test)

    # подсчет метрики на тестовой выборке
    mse.append(mean_squared_error(y_test, y_pred))
    mae.append(mean_absolute_error(y_test, y_pred))
    mape.append(mean_absolute_percentage_error(y_test, y_pred))

In [16]:
mse, mae, mape

([2992.5812293010176,
  2794.8881014302633,
  3010.6150587576667,
  2937.8713184385324,
  3424.259334298692,
  2471.174580381414],
 [41.97492114949366,
  41.751319833264006,
  45.09846060863129,
  44.44205567958728,
  46.17358500370479,
  41.13279719843048],
 [32.494313191950184,
  38.73424853778743,
  41.31391834737482,
  38.98916836082025,
  38.0453316585361,
  40.64797082454018])

Проверим меняются ли метрики при замене разбиения данных

In [20]:
mse = []
mae = []
mape = []

sizes = np.array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])

for size in sizes:

    X_train, X_test = train_test_split(X, test_size=size, random_state = 4)
    y_train, y_test = train_test_split(y, test_size=size, random_state = 4)

    model = LinearRegression(fit_intercept=True)  # объявляем модель
    model.fit(X_train, y_train)  # обучаем


    # предсказание модели
    y_pred = model.predict(X_test)

    # подсчет метрики на тестовой выборке
    mse.append(mean_squared_error(y_test, y_pred))
    mae.append(mean_absolute_error(y_test, y_pred))
    mape.append(mean_absolute_percentage_error(y_test, y_pred))

In [18]:
mse, mae, mape

([3572.0491569032483,
  3507.3634069591417,
  3298.546380516042,
  3200.098153478155,
  3122.209069165025,
  3202.8005592376376,
  3204.166191374347,
  4132.135823580124,
  4674.746512497638],
 [49.69620150007066,
  47.96890732842747,
  46.485917678326345,
  44.87572490372343,
  44.06753428454939,
  44.54355381037322,
  44.837715068816735,
  48.203571508400486,
  51.03482469907298],
 [45.89654142652991,
  40.60854800727221,
  42.33338562873135,
  40.07764977730673,
  38.055473342838944,
  37.93959949243264,
  37.13939081303273,
  39.28280675808343,
  40.459212612035614])

**Вывод:** В среднем на протяжении всей работы MAPE была на уровне 35-40%. Можем видеть, что при замене random_state (по факту сида) ничего не меняется. Также смотря на долю тренировочных данных (size), кажется, что ошибка имеет экстремум, где-то в центре, хотя это требует дополнительных исследований.